### PR TITLE
Tabular: Added transform_features function to TabularPredictor

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*
#249, #428

*Description of changes:*

Added transform_features function to TabularPredictor

Now users can utilize this function to get feature transformations on their datasets to use with their own custom models outside of AutoGluon, even enabling users to extend the stack ensemble by receiving the out-of-fold prediction features of the AutoGluon ensemble.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
Code Example of new functionality:


```
import pandas as pd
import sklearn.datasets

from autogluon.utils.tabular.ml.utils import generate_train_test_split
from autogluon import TabularPrediction as task

LABEL = 'target'


if __name__ == '__main__':
    cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
    X = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
    X[LABEL] = cancer_data[LABEL]

    X_train, X_test, y_train, y_test = generate_train_test_split(X=X.drop(LABEL, axis=1), y=X[LABEL], problem_type='binary', test_size=0.2)

    X = X_train
    X[LABEL] = y_train
    X_test[LABEL] = y_test

    predictor_stack = task.fit(
        train_data=X, label=LABEL,
        hyperparameters={
            'GBM': {'num_boost_round': 10},
            'CAT': {'iterations': 10},
        },
        num_bagging_folds=2, stack_ensemble_levels=1,
        num_bagging_sets=1,
    )

    a = predictor_stack.transform_features(X_test)
    b = predictor_stack.transform_features(X_test, base_models=['LightGBMClassifier_STACKER_l0', 'weighted_ensemble_k0_l2'])
    c = predictor_stack.transform_features(X_test, model='weighted_ensemble_k0_l2')
    d = predictor_stack.transform_features()
    e = predictor_stack.transform_features(base_models=['LightGBMClassifier_STACKER_l0', 'weighted_ensemble_k0_l2'])
    f = predictor_stack.transform_features(model='weighted_ensemble_k0_l2')

    print('#############')
    print(a[:3])
    print(b[:3])
    print(c[:3])
    print(d[:3])
    print(e[:3])
    print(f[:3])
    print('#############')
```

Code Output:

```
#############
     mean radius  mean texture  ...  worst symmetry  worst fractal dimension
256        19.55         28.77  ...          0.2818                  0.10050
428        11.13         16.62  ...          0.2383                  0.07083
501        13.82         24.49  ...          0.3651                  0.11830
[3 rows x 30 columns]
   LightGBMClassifier_STACKER_l0  ...  worst fractal dimension
0                       0.256184  ...                      NaN
1                       0.850992  ...                  0.08902
2                       0.373475  ...                      NaN
[3 rows x 32 columns]
   CatboostClassifier_STACKER_l1  ...  worst fractal dimension
0                       0.243614  ...                      NaN
1                       0.769152  ...                  0.08902
2                       0.336421  ...                      NaN
[3 rows x 31 columns]
     mean radius  mean texture  ...  worst symmetry  worst fractal dimension
546        10.32         16.35  ...          0.2681                  0.07399
432        20.18         19.54  ...          0.3032                  0.08075
174        10.66         15.15  ...          0.2710                  0.06164
[3 rows x 30 columns]
     LightGBMClassifier_STACKER_l0  ...  worst fractal dimension
546                       0.850150  ...                  0.07399
432                       0.250364  ...                  0.08075
174                       0.847202  ...                  0.06164
[3 rows x 32 columns]
     CatboostClassifier_STACKER_l1  ...  worst fractal dimension
546                       0.670516  ...                  0.07399
432                       0.167280  ...                  0.08075
174                       0.690598  ...                  0.06164
[3 rows x 31 columns]
#############
```

**Comment by szha** [comment]
Job PR-431-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-431/1/index.html

**Comment by szha** [comment]
Job PR-431-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-431/2/index.html

File: autogluon/task/tabular_prediction/predictor.py [hunk 1]
```diff
@@ -341,6 +341,64 @@ def fit_summary(self, verbosity=3):
...
+        When `dataset=None`, `base_models=[{best_model}], and bagging was enabled during fit():
+            This returns the out-of-fold predictions of the best model, which can be used as training input to a custom user stacker model.
+
+        Parameters
+        ----------
+        dataset : str or :class:`TabularDataset` or `pandas.DataFrame` (optional)
+            The dataset to apply feature transformation to.
+            This dataset does not require the label column.
+            If str is passed, `dataset` will be loaded using the str value as the file path.
+            If not specified, the original dataset used during fit() will be used if `cache_data=True`. Otherwise, an exception will be raised.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:358]: if `cache_data=True`
->
if fit() was previously called with `cache_data=True`

File: autogluon/task/tabular_prediction/predictor.py [hunk 2]
```diff
@@ -341,6 +341,64 @@ def fit_summary(self, verbosity=3):
...
         return results
 
+    def transform_features(self, dataset=None, model=None, base_models=None):
+        """
+        Transforms dataset features through the AutoGluon feature generator.
+        This is useful to gain an understanding of how AutoGluon interprets the dataset features.
+        The output of this function can be used to train further models, even outside of AutoGluon.
+        When `dataset=None`, `base_models=[{best_model}], and bagging was enabled during fit():
+            This returns the out-of-fold predictions of the best model, which can be used as training input to a custom user stacker model.
+
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:351]: Include small example in the documentation? 
This function is a bit opaque without an example.

File: autogluon/task/tabular_prediction/predictor.py [hunk 3]
```diff
@@ -341,6 +341,64 @@ def fit_summary(self, verbosity=3):
             print("*** End of fit() summary ***")
         return results
 
+    def transform_features(self, dataset=None, model=None, base_models=None):
+        """
+        Transforms dataset features through the AutoGluon feature generator.
+        This is useful to gain an understanding of how AutoGluon interprets the dataset features.
```
Comment (jwmueller) [inline autogluon/task/tabular_prediction/predictor.py:347]: Should clarify that: Individual AutoGluon models like the neural network may apply additional feature transformations that are not reflected in this method, which only applies universal transforms employed by all AutoGluon models.

File: autogluon/task/tabular_prediction/predictor.py (hunk 1 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py]: Added

File: autogluon/task/tabular_prediction/predictor.py (hunk 3 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py:347]: Added

File: autogluon/task/tabular_prediction/predictor.py (hunk 2 above)
Comment (Innixma) [reply autogluon/task/tabular_prediction/predictor.py:353]: Added

**Comment by szha** [comment]
Job PR-431-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-431/3/index.html

**Pull request merged**